ext/drstatecmp/drstatecmp.c

```diff
@@ -0,0 +1,606 @@
+     * bb. No change needed since data->golden_bb_copy already points to the correct copy.
+     */
+    return DR_EMIT_DEFAULT;
+}
+
+/****************************************************************************
+ * INSTRUMENTATION INSERTION PHASE
+ *
+ * Instrumentation insertion pass is used to maintain the user_data (created in the
+ * app2app phase) for the post-instrumentation phase.
+ *
+ */
+
+static dr_emit_flags_t
+drstatecmp_insert_phase(void *drcontext, void *tag, instrlist_t *bb, instr_t *instr,
+                        bool for_trace, bool translating, void *user_data)
+{
+    return DR_EMIT_DEFAULT;
+}
+
+/****************************************************************************
+ * POST-INSTRUMENTATION PHASE
+ *
+ * In this phase, drstatecmp inserts all the necessary state comparisons.
+ */
+
+static void
+drstatecmp_duplicate_bb(void *drcontext, instrlist_t *bb, drstatecmp_user_data_t *data,
+                        drstatecmp_dup_labels_t *labels)
+{
+    /* Duplication process.
+     * Consider the following example bb:
+     *   instr1
+     *   meta_instr
+     *   instr2
+     *   term_instr
+     *
+     * In this stage, we just duplicate the bb (except for its terminating
+     * instruction and meta instructions) and add special labels to the original and
+     * duplicated blocks. Note that there might be no term_instr (no control
+     * transfer instruction) and the bb just falls-through. Also note that we use for the
+     * bb copy the golden_copy that was determined in the analysis phase. This golden copy
+     * contains the app instructions of the bb without any of the instrumentation. It
+     * might also contain app2app changes if emulation was required.
+     *
+     * The example bb is transformed, in this stage, as follows:
+     * ORIG_BB:
+     *   instr1
+     *   meta_instr
+     *   instr2
+     *
+     * COPY_BB:
+     *   instr1
+     *   instr2
+     *
+     * TERM:
+     *   term_instr
+     *
+     */
+
+    /* Get an instrumentation-free copy of the bb which is the golden copy kept
+     * in the user_data*/
+    instrlist_t *copy_bb = data->golden_bb_copy;
+
+    /* Create and insert the labels. */
+    labels->orig_bb_start = drstatecmp_insert_label(drcontext, bb, instrlist_first(bb),
+                                                    DRSTATECMP_LABEL_ORIG_BB,
+                                                    /*preinsert=*/true);
+    labels->copy_bb_start =
+        drstatecmp_insert_label(drcontext, copy_bb, instrlist_first(copy_bb),
+                                DRSTATECMP_LABEL_COPY_BB, /*preinsert=*/true);
+    /* Insert the TERM label before the terminating instruction or after the
+     * last instruction if the bb falls through.
+     */
+    instr_t *term_inst_copy_bb = instrlist_last_app(copy_bb);
+    bool preinsert = true;
+    if (!instr_is_cti(term_inst_copy_bb) && !instr_is_return(term_inst_copy_bb))
```

Comment:
"cti" should include return

Reply:
You are right. Double-checked it and indeed instr_is_cti() checks for instr_is_mbr_arch() which is a superset of instr_is_return().